### PR TITLE
[batch] fix bugs in worker

### PR DESCRIPTION
**Pull request by jigold**
This fixes two bugs:
1. The container logs weren't being cached. This made the logs "disappear" for previous tasks while the job was still running. FYI @konradjk 

2. My job got stuck in "running" even though the job was deleted from the worker because writing the status to GCS timed out and we didn't actually mark the job complete. I'm not sure if we should always try to retry writing the status rather than failing on non-transient errors.

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.6/http/client.py", line 1354, in getresponse
    response.begin()
  File "/usr/local/lib/python3.6/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.6/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/local/lib/python3.6/ssl.py", line 1012, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/local/lib/python3.6/ssl.py", line 874, in read
    return self._sslobj.read(len, buffer)
  File "/usr/local/lib/python3.6/ssl.py", line 631, in read
    v = self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/retry.py", line 368, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.6/site-packages/urllib3/packages/six.py", line 686, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 386, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 306, in _raise_timeout
    raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='www.googleapis.com', port=443): Read timed out. (read timeout=60)

During handling of the above exception, another exception occurred
result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/site-packages/hailtop/utils/utils.py", line 35, in <lambda>
    thread_pool, lambda: fun(*args, **kwargs))
  File "/usr/local/lib/python3.6/site-packages/batch/google_storage.py", line 65, in _write_gs_file
    f.upload_from_string(string, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 1257, in upload_from_string
    predefined_acl=predefined_acl,
  File "/usr/local/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 1157, in upload_from_file
    client, file_obj, content_type, size, num_retries, predefined_acl
  File "/usr/local/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 1063, in _do_upload
    client, stream, content_type, size, num_retries, predefined_acl
  File "/usr/local/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 857, in _do_multipart_upload
    response = upload.transmit(transport, data, object_metadata, content_type)
  File "/usr/local/lib/python3.6/site-packages/google/resumable_media/requests/upload.py", line 106, in transmit
    retry_strategy=self._retry_strategy,
  File "/usr/local/lib/python3.6/site-packages/google/resumable_media/requests/_helpers.py", line 136, in http_request
    return _helpers.wait_and_retry(func, RequestsMixin._get_status_code, retry_strategy)
  File "/usr/local/lib/python3.6/site-packages/google/resumable_media/_helpers.py", line 150, in wait_and_retry
    response = func()
  File "/usr/local/lib/python3.6/site-packages/google/auth/transport/requests.py", line 317, in request
    **kwargs
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='www.googleapis.com', port=443): Read timed out. (read timeout=60)
```

### CODE REVIEW

File: hail/python/hailtop/utils/utils.py [hunk 1]
```diff
@@ -205,6 +211,10 @@ def is_transient_error(e):
                 e.errno == errno.EHOSTUNREACH or
                 e.errno == errno.ECONNRESET):
             return True
+    elif isinstance(e, urllib3.exceptions.ReadTimeoutError):
```
Comment (danking) [inline hail/python/hailtop/utils/utils.py]: need to catch:
```
requests.exceptions.ReadTimeout
```

File: batch/batch/google_storage.py [hunk 2]
```diff
@@ -34,19 +34,24 @@ def __init__(self, blocking_pool, *, project=None, credentials=None):
         self._wrapped_delete_gs_files = self._wrap_network_call(GCS._delete_gs_files)
 
     async def write_gs_file(self, uri, string, *args, **kwargs):
-        return await self._wrapped_write_gs_file(self, uri, string, *args, **kwargs)
+        return await retry_transient_errors(self._wrapped_write_gs_file,
```
Comment (danking) [inline batch/batch/google_storage.py]: can you run a test where you write to the same uri the same string twice? I'm worried that we need to specify an overwrite command or something.

File: batch/batch/google_storage.py [hunk 3]
```diff
@@ -34,19 +34,24 @@ def __init__(self, blocking_pool, *, project=None, credentials=None):
...
-        return await self._wrapped_read_binary_gs_file(self, uri, *args, **kwargs)
+        return await retry_transient_errors(self._wrapped_read_binary_gs_file,
+                                            self, uri, *args, **kwargs)
 
     async def delete_gs_file(self, uri):
-        return await self._wrapped_delete_gs_file(self, uri)
+        return await retry_transient_errors(self._wrapped_delete_gs_file,
+                                            self, uri)
 
     async def delete_gs_files(self, uri_prefix):
```
Comment (danking) [inline batch/batch/google_storage.py]: can you verify that deleting a URI and a URI prefix that is already deleted does not error?

File: batch/batch/worker.py [hunk 4]
```diff
@@ -287,10 +287,12 @@ def step(self, name, **kwargs):
             log.info(f'{self}: container status {self.container_status}')
 
             async with self.step('uploading_log'):
+                container_log = await self.get_container_log()
+                self.log = container_log
```
Comment (danking) [inline batch/batch/worker.py]: let's move this in get_container_log

File: batch/batch/worker.py [hunk 5]
```diff
@@ -846,6 +849,8 @@ def __init__(self):
     async def post_job_complete(self, job, run_duration):
         try:
             await self.post_job_complete_1(job, run_duration)
+        except Exception:
+            log.exception(f'error while marking {job} complete')
```
Comment (danking) [inline batch/batch/worker.py]: let's make this a separate PR

File: batch/batch/google_storage.py (hunk 2 above)
Comment (jigold) [reply batch/batch/google_storage.py]: It worked fine.

File: batch/batch/google_storage.py (hunk 3 above)
Comment (jigold) [reply batch/batch/google_storage.py]: It errors in the first case, but I added a catch in the second case in case files get deleted after listing the files in the directory.

File: batch/batch/google_storage.py [hunk 6]
```diff
@@ -84,10 +89,16 @@ def _delete_gs_files(self, uri_prefix):
...
-            blob.delete()
+            try:
+                blob.delete()
+            except google.api_core.exceptions.NotFound:
+                continue
 
     def _delete_gs_file(self, uri):
         bucket, path = GCS._parse_uri(uri)
         bucket = self.gcs_client.bucket(bucket)
         f = bucket.blob(path)
```
Comment (danking) [inline batch/batch/google_storage.py]: This should be inside the try as well. Presumably this could fail if the path does not exist.

Comment (jigold) [reply batch/batch/google_storage.py]: I don't think so. I think it's just initializing an object in Python. Doesn't actually do any API requests. https://googleapis.dev/python/storage/latest/buckets.html#google.cloud.storage.bucket.Bucket.blob

Comment (danking) [reply batch/batch/google_storage.py]: great
